text/0047-unspecified-semantics.md

```diff
@@ -0,0 +1,178 @@
+# Make `==` and `in` on `Unspecified` to be errors
+
+## Related issues and PRs
+
+- Reference Issues: [cedar#217](https://github.com/cedar-policy/cedar/issues/217)
+- Implementation PR(s):
+
+## Timeline
+
+- Started: 2024-02-05
+
+## Summary
+
+The `==` and `in` operations will always produce errors when applied to
+`Unspecified` entities. This replaces the current behavior, which produces
+errors in some cases, `true` in other cases, and `false` in other cases.
```

Comment:
Then: It's now possible for expressions in the scope to have an error. I imagine this affects how we do policy slicing. Have you thought that through?

Reply:
Good catch, I have not thought this through

Reply:
The soundness of slicing hinges on the fact that scope constraints cannot error.  We should not violate this invariant without a very good reason.

The current behavior of `Unspecified` is consistent with what you would get if you declared a special type with no attributes that is distinct from all other types, and assigned your principal to a dummy value of this special type.  I don't think the semantics of this is confusing or hard to explain.

If we're going to make a breaking change, then I would vote for completely dropping the support for `Unspecified`.

Reply:
A naive question:  why do we need to have any special handling of `Unspecified`?  Can we drop it from validation, PE, etc., and just leave runtime APIs that that implement `Unspecified` as literally a value of type with no attributes, with the value generated by the Rust code? 

If we choose a type name that is not a valid Cedar identifier, that type is guaranteed to be distinct from all other types.

Reply:
Basically, can we sort of have our cake (runtime support for `Unspecified`) and eat it too (no special handling in the tooling)?

Reply:
Admittedly, it's disappointing to have runtime support but not tooling support.  So maybe the cleanest course of action is to drop `Unspecified`.

Reply:
We originally implemented `Unspecified` basically that way.  However, rather than an illegal entity-type identifier, we thought it much clearer and error-proof to use an `Option` type.  It was thought that forcing users of `EntityType` to explicitly consider the `Unspecified` case was a virtue and would lead to fewer bugs, particularly in validation/PE/etc.

Reply:
Consider that the validator has to consider the possibility that values exist of type `Unspecified`, even though the `Unspecified` type is not declared in the schema.  With the `Option` or explicitly-enumerated type, the validator basically cannot forget this.  If `Unspecified` was simply implemented as a value with illegal entity type, the validator could easily forget to handle the case for that entity type.

Reply:
Do we have a way to estimate the impact of dropping `Unspecified`?

Reply:
> Consider that the validator has to consider the possibility that values exist of type `Unspecified`, even though the `Unspecified` type is not declared in the schema. With the `Option` or explicitly-enumerated type, the validator basically cannot forget this. If `Unspecified` was simply implemented as a value with illegal entity type, the validator could easily forget to handle the case for that entity type.

My naive thought is that we would automatically extend _every_ user schema with the special `Unspecified` type(s) and then just run normal validation against this extended schema.

Where does this naive approach break down?

Reply:
Why do we need to do that? Why shouldn't users define their own "unspecified" types when they want to make up a value to pass it in? Then they can include this type in the schema, which will make clear it's a member of nothing and has no attributes?

Maybe this conversation belongs in a new RFC ...

Reply:
I think the proposals in this thread (switch to treating unspecified entities as entities with a unique type for evaluation and validation) are viable. This is originally how we did it, but we then switched to encoding this information at the type level because we thought this would be less error prone (hah!).

I worry that making this change in the validator will cause us to change precision in some cases, which could lead to breaking changes. But I'd have to look at the code more carefully to be sure.

Reply:
Right, the above proposal is a mechanical way to simulate exactly that:  users defining their own "unspecified" types in the schema and having actions refer to those unspecified types.

I also think it's a cleaner design for users to do this.

The proposal is a way to avoid making a breaking change if we decide that getting rid of `Unspecified` (preferred) is not an option at this time.

Reply:
I don't immediately see a problem with handling unspecified as an implicit type in the schema if the semantics actually match that.